pkg/tests/tasks/security/certificate/external_ca_test.go

```diff
@@ -15,109 +15,193 @@
 package certificate
 
 import (
+	"bufio"
+	"crypto/x509"
+	"encoding/base64"
 	"fmt"
+	"io"
 	"os"
+	"strings"
 	"testing"
 
 	"github.com/maistra/maistra-test-tool/pkg/app"
-	"github.com/maistra/maistra-test-tool/pkg/util"
 	"github.com/maistra/maistra-test-tool/pkg/util/check/assert"
 	"github.com/maistra/maistra-test-tool/pkg/util/curl"
 	"github.com/maistra/maistra-test-tool/pkg/util/env"
 	"github.com/maistra/maistra-test-tool/pkg/util/oc"
 	"github.com/maistra/maistra-test-tool/pkg/util/pod"
 	"github.com/maistra/maistra-test-tool/pkg/util/retry"
-	"github.com/maistra/maistra-test-tool/pkg/util/shell"
 	"github.com/maistra/maistra-test-tool/pkg/util/test"
 )
 
 func TestExternalCertificate(t *testing.T) {
 	test.NewTest(t).Id("T17").Groups(test.Full, test.ARM, test.InterOp).Run(func(t test.TestHelper) {
 		const ns = "bookinfo"
 
-		var tmpDir = ""
-
 		t.Cleanup(func() {
-			if tmpDir != "" {
-				os.RemoveAll(tmpDir)
-			}
-
+			t.Logf("Recreate namespace %s", ns)
 			oc.RecreateNamespace(t, ns, meshNamespace)
 		})
 
 		t.LogStep("Uninstall existing SMCP")
 		oc.RecreateNamespace(t, meshNamespace)
 
-		t.LogStep("Create cacerts Secret")
+		t.LogStep("Create cacerts secret")
 		oc.CreateGenericSecretFromFiles(t, meshNamespace, "cacerts",
 			"ca-cert.pem="+sampleCACert,
 			"ca-key.pem="+sampleCAKey,
 			"root-cert.pem="+sampleCARoot,
 			"cert-chain.pem="+sampleCAChain)
+		rootCert := readPemCertificatesFromFile(t, sampleCARoot)[0]
+		chainCerts := readPemCertificatesFromFile(t, sampleCAChain)
 
-		t.LogStep("Apply SMCP to configure certificate authority to use cacerts Secret")
+		t.LogStep("Apply SMCP to configure certificate authority to use cacerts secret")
 		oc.ApplyString(t, meshNamespace, createSMCPWithCustomCert(smcpName, ns))
 		oc.WaitSMCPReady(t, meshNamespace, smcpName)
 
 		t.LogStep("Install bookinfo")
 		app.InstallAndWaitReady(t, app.BookinfoWithMTLS(ns))
 
-		t.LogStep("Checking response from productpage.")
+		t.LogStep("Wait for response from productpage app")
 		retry.UntilSuccess(t, func(t test.TestHelper) {
 			curl.Request(t,
 				app.BookinfoProductPageURL(t, meshNamespace), nil,
 				assert.ResponseStatus(200))
 		})
 
-		if env.GetArch() == "p" || env.GetArch() == "z" {
-			t.Log("NOTE: Not checking certificates, because test is running in P or Z environment")
-			return
-		}
+		t.LogStep("Retrieve certificates from bookinfo details service")
+		opensslOutput := oc.Exec(t, pod.MatchingSelector("app=productpage", ns), "istio-proxy",
+			`openssl s_client -showcerts -connect details:9080 || true`) // the "|| true" is needed until oc.Exec can ignore return codes
```

Comment:
Because `readPemCertificatesFromText` will fail if no certificates could be read, I believe it should be sufficient to only retry only this section of attempting to fetch and then parse the certificates.

Reply:
I believe you could also get rid of the `|| true`, right?

Reply:
The `|| true` is there because oc/kubectl will pass along the return code from the remote openssl command (or at the very least fail if openssl fails) and openssl is expected to fail here. I would need a way to tell oc.Exec that I want to verify the return code myself or at least to ignore the return code.

Additional info if interested:
The `s_client` subcommand, used here for returning the certificate data, actually also dumps other debug info and attempts a test verification. This does not seem to be configurable without also silencing the certificate data, and there does not seem to be a more direct way to just dump the certs and do nothing else. We aren't running the command for the purpose of verifying, so adding additional complexity to the test to make sure the remote openssl is pointed at the correct root cert for a successful verification we ultimately will ignore seems like a worse workaround to me than the `|| true`.

Tangent:
Another issue with oc.Exec this points out is that I believe the `|| true` is actually interpreted as part of the local command line, not the remote one. So it's not `oc exec "remote command || true"` (which is what I would expect), it's `oc exec "remote command" || true`.

Reply:
Yeah, but the openssl command only fails on the first (few?) attempts. I removed the `|| true` and the command was successful on the second attempt.

And I think you're right about the `|| true` being executed locally instead of in the pod. We should fix that, as I'm sure someone will eventually waste a lot of time because of this unexpected behavior.